Automattic/wp-calypso
# Checkout: Add immediate redirect on pre-filled receipt ID to pending page

#### Proposed Changes

The post-checkout "pending" page is visited by most payment methods ([and more](https://github.com/Automattic/wp-calypso/pull/65273) [soon](https://github.com/Automattic/wp-calypso/pull/65327)). It gets an order ID and then polls for the status of that order until the order is complete (or failed). Once complete, it redirects the user to the URL in the `redirectTo` query parameter.

For most transactions, the order ID is added to the "pending" page URL by either an API endpoint (eg: for redirect payment methods) or directly when the "pending" page URL is created. If the transaction is already complete by the time the URL loads, it will redirect almost instantly since the order endpoint will confirm the success. However, there are some transactions that do not produce an order: free purchases.

Fortunately, by the time a free purchase transaction endpoint request completes, it is always successful (a failure would not complete). So if such purchases use the "pending" page, they really only need to have it go directly to the `redirectTo` query parameter, but this won't work because without an order ID, the "pending" page has no way to confirm that the order was successful.

In this PR, we modify the "pending" page URL to add a new query param, `receiptId`. If set to a number, the "pending" page will know that the transaction has completed and can ignore the order ID (or lack of one).

Because it's relatively easy to provide a receipt ID in the URL, we make it easy to trigger the open redirect already present in the "pending" page (on success, it redirects to whatever is in the `redirectTo` query string). Therefore, this PR also changes the pending page so that it will only redirect to known hosts, similar to https://github.com/Automattic/wp-calypso/pull/54414.

This is needed by https://github.com/Automattic/wp-calypso/pull/65273 and is part of implementing https://github.com/Automattic/wp-calypso/issues/53356.

#### Testing Instructions

Nothing uses this functionality currently. To test it, try a free purchase (eg: buying a domain using a domain credit) using https://github.com/Automattic/wp-calypso/issues/65273 and verify that you end up on a thank-you page with the receipt ID in the URL rather than a generic thank-you page that has `unknown` in the URL.

To that end, this PR adds unit tests for all the possible redirect cases.
